### ISSUE TITLE
Render react component inside winbox?

### ISSUE THREAD

**Issue by Kielx** [nextapps-de/winbox#36]
Is it possible to render React component inside winbox?

I tried to use a component inside innerHtml, html, body, and mount options but to no avail. Is it possible to render a component inside the winbox element or am I out of luck? Code below renders nothing

```javascript
        <a
          style={{ cursor: "pointer" }}
          onClick={() => {
            new WinBox({
              title: "About Me",
              modal: false,
              x: "20%",
              y: "20%",
              onfocus: function () {
                this.setBackground("#00aa00")
              },
              onblur: function () {
                this.setBackground("#777")
              },
              innerHtml: <About></About>,
            })
          }}
        >
          /About
        </a>
```

**Comment by fzammetti** [comment]
Ooh!  I know this one! :)

I'm building an app that is sort of a "web desktop" concept, so I have lots of little apps that launch in WinBox windows.  Each of those apps is built with React, so I'm doing exactly what you're asking about.

The trick is you have to build the window first, THEN build the root React component in it.  I don't think there's a way to do it like you're showing here (I could be wrong, but I know I couldn't figure it out).

So, what I do is this:

```
const rootComponent = await import("app_root_component");
const win = new WinBox({ title : "My app", width : 300, height : 500 });
ReactDOM.render(React.createElement(rootComponent), win.body);
```

app_root_component is a jsx file.  Because my apps can be loaded on-the-fly by the user ala a desktop GUI, that's the reason I'm importing the module that way.  If you know before-hand what modules you need then you can simply import them like you would any other module and skip that part.  But, that's the basic idea: construct the window, then construct the React component in it's body, and you're off to the races.

**Comment by Kielx** [comment]
Hey Frank! Thank You very much for your answer. It helped me, but I'm still struggling with the implementation of your idea. 
I'm using mostly functional components in my app and would like to create winbox window as a result of onClick event.
However, your implementation uses `ReactDOM.render` - which, as far as I know, should be only used when mounting react app. Therefore I'm unable to create winbox instances dynamically. Your solution works for me - true, but winbox instance is created every time while I would like it to be created only as a result of an event. 

In other words - Is there any other way to append jsx to win.body other than in `ReactDOM.render(React.createElement(rootComponent), win.body)`?

**Comment by fzammetti** [comment]
You can use ReactDOM.render() at any time to render a component.  In fact, if memory serves, that's what React uses under the covers when creating a component tree.  True, you almost never would call it explicitly if you are writing a pure React app, but in this case, you're in a sense not with WinBox in the mix.  The second argument to it tells it what existing DOM node to build the React content under, which is the key, since it's the body of the WinBox window here.

In my app, I have a sidebar provided by the Material-UI library that forms the "base UI" of the app, so to speak.  When you click on item in the list on it, it creates a WinBox window, and then populates it as described with the React content, so it sounds like I'm doing exactly what you're describing.  My onClick event on the list item calls a function that, in essence, does what I posted earlier (there's a lot more app-specific stuff it does, but none of that would matter to you).

Are you by chance executing the code I posted outside the onClick handler, like maybe on page load or something?  If so, then I think that's the disconnect... just do it in your onClick handler instead.  Now, if you want to ensure there's only ever a single window from those click events, you'll have to add some code for that (I do that in my app too: I keep a reference to the created window and only build a new one if it's null).

Hope that helps!

**Comment by Kielx** [comment]
My sincere apologies. Indeed it was my mistake I made while importing a component. 
Thank You once again for your excellent help and detailed descriptions! Everything works as it should now!

**Comment by fzammetti** [comment]
No need to apologize, we're all just folk here, and we all make simple mistakes sometimes :)  Glad you got it working!

**Comment by wobsoriano** [comment]
For Vue users, you can check this wrapper component https://github.com/wobsoriano/vue-winbox.

It lets you use Vue components inside the WinBox body.

**Comment by RickoNoNo3** [comment]
Now [an entirely React-controlled component](https://github.com/RickoNoNo3/react-winbox) for WinBox is available. I want it to be better so if someone can try to use it and issue some problems, I will thanks a lot.
